### PR TITLE
code cleanup: Use libvmi builder pattern instead of NewRandomVMI function

### PR DESCRIPTION
**Pull request by Barakmor1**
Add options to libvmi in order to use libvmi builder pattern in functional tests instead of NewRandomVMI functions.
**What this PR does / why we need it**:
Using libvmi instead of NewRandomVMI functions.
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: tests/libvmi/vmi.go
Comment (iholder-redhat) [inline tests/libvmi/vmi.go:150]: The `secureBoot` option is a leftover and should be removed entirely imo

File: tests/libvmi/vmi.go
Comment (iholder-redhat) [inline tests/libvmi/vmi.go:161]: Please change `pointer.BoolPtr` -> `pointer.Bool`
Also, this needs to always be true (as the function name is WithSerialBios)

File: tests/vmi_configuration_test.go
Comment (iholder-redhat) [inline tests/vmi_configuration_test.go:537]: `var interfaceDeviceWithMasqueradeBinding = libvmi.InterfaceDeviceWithMasqueradeBinding()`
->
`interfaceDeviceWithMasqueradeBinding := libvmi.InterfaceDeviceWithMasqueradeBinding()`

File: tests/vmi_configuration_test.go
Comment (iholder-redhat) [inline tests/vmi_configuration_test.go:547]: Personally I don't like this very much.
I would calculate this before:

```go
resourceMemory := "256Mi"
if !checks.IsARM64(testsuite.Arch) {
    resourceMemory = "128Mi"
}

...

vmi := libvmi.New(
    resourceMemory,
    ...
)
```

Comment (iholder-redhat) [reply tests/vmi_configuration_test.go:547]: But, thinking about this again, I'm not sure why it's needed at all.
When we convert a VMI into a target pod, we [take care](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/services/template.go#L1479) of such memory allocations.

I think we can remove this altogether, or add this to one of those:
- renderLaunchManifest
- libvmi's New()
- libvmi's constructors (NewFedora, NewCirros, etc).

This can be done in a follow-up PR, but I would at least simply do:
```go
if !checks.IsARM64(testsuite.Arch) {
    resourceMemory = "128Mi"
}
```

Since the other case (where it is an arm64 arch) is taken care of inside renderLaunchManifest

File: tests/usbredir_test.go
Comment (iholder-redhat) [inline tests/usbredir_test.go:182]: In which scenarios this is needed?
Why not include it within `libvmi.New()`?

File: tests/usbredir_test.go
Comment (iholder-redhat) [inline tests/usbredir_test.go:180]: nice!

File: tests/usbredir_test.go
Comment (iholder-redhat) [inline tests/usbredir_test.go:192]: We can also check for errors here and avoid returning an error from this func

**Review by iholder-redhat — COMMENTED** [review]
Great work @Barakmor1!

File: tests/libvmi/vmi.go
Comment (Barakmor1) [reply tests/libvmi/vmi.go]: Done.

File: tests/libvmi/vmi.go
Comment (Barakmor1) [reply tests/libvmi/vmi.go]: Done.

File: tests/vmi_configuration_test.go
Comment (Barakmor1) [reply tests/vmi_configuration_test.go]: Done.

File: tests/usbredir_test.go
Comment (Barakmor1) [reply tests/usbredir_test.go:180]: As you said it is out of the scope of this PR if it's indeed need to be done.

File: tests/usbredir_test.go
Comment (Barakmor1) [reply tests/usbredir_test.go]: Done.

File: tests/vmi_configuration_test.go
Comment (iholder-redhat) [inline tests/vmi_configuration_test.go:542]: In a future PR we should remove such things from tests entirely and handle them within renderLaunchManifest.
FYI @maiqueb, it might be related to the PR you're working on

**Review by iholder-redhat — COMMENTED** [review]
Thanks @Barakmor1! Great stuff

/lgtm

**Comment by iholder-redhat** [comment]
@Barakmor1 can you please change the PR title to something like `Add libvmi options for CPU resources, SerialBIOS and ClientPassthrough`?

File: tests/vmi_configuration_test.go
Comment (jean-edouard) [inline tests/vmi_configuration_test.go:1862]: This is not enough memory for basically VMI, is it on purpose so that it will fail to start?

**Review by jean-edouard — COMMENTED** [review]
Looks good overall but please see in-line comment.

File: tests/vmi_configuration_test.go
Comment (Barakmor1) [reply tests/vmi_configuration_test.go:1862]: Hey Jed,
Thanks for reviewing
If i understood correctly NewRandomVMI generate diskless vmi so it doesn't need memory.
The reason for the 1Mi is because virt-api doesn't let creating vmi without memory.
The purpose of this is to be more resources efficient.

Comment (jean-edouard) [reply tests/vmi_configuration_test.go:1862]: Diskless VMs still need memory, for things like the firmware. I guess it's possible that the emulated BIOS manages to run on such a small amount of memory (EFI definitely wouldn't), but it still feels extremely tight. Can we do something like 32Mi? It remains quite insignificant for any CI infrastructure, and it would ensure future updates don't break tests.

Comment (xpivarc) [reply tests/vmi_configuration_test.go:1862]: Just note: EFI needs a hundred or so if I remember correctly.

File: tests/libvmi/vmi.go
Comment (xpivarc) [inline tests/libvmi/vmi.go:150]: Same as below

File: tests/vmi_configuration_test.go
Comment (iholder-redhat) [reply tests/vmi_configuration_test.go:1862]: Please bear in mind that in `template.go`'s `renderLaunchManifest()`, where we convert a VMI manifest into a target pod, we take care (and should take care) of the required memory allocations in order for the desired spec to be possible. It is not the user's responsibility. Any memory that the user requests is accumulated into the memory we allocate automatically.

I'm not sure about EFI specifically, but if we don't automatically add memory for it in `renderLaunchManifest()`, then we should :)

Comment (jean-edouard) [reply tests/vmi_configuration_test.go:1862]: @itamarh I believe what you are referring to is the memory allocation for the compute container.
What I'm talking about here is the actual amount of RAM the VM sees, which I don't think we mess with, or do we?

Comment (iholder-redhat) [reply tests/vmi_configuration_test.go:1862]: @jean-edouard and I were talking offline. Here's a summary of our conclusions.

Currently it's the user's responsibility to allocate the memory needed for the guest, so currently 1MB wouldn't be enough. But this behavior is wrong. The user (or the tests for that matter) shouldn't worry about such low-level-implementation-detail considerations - we should.

We think the solution would be to enforce a minimum RAM amount, and set it as a default.

Until then we can start with `32Mi`, which @jean-edouard suggested as a good starting point.

Comment (Barakmor1) [reply tests/vmi_configuration_test.go]: Done.
Thanks!

File: tests/libvmi/vmi.go
Comment (Barakmor1) [reply tests/libvmi/vmi.go:165]: What is the value of making it local if we will probably use it again?

Comment (xpivarc) [reply tests/libvmi/vmi.go:165]: @iholder-redhat  
I would argue the same for `WithSEV`. It should be in `sev.go` where it would be pretty self-contained.

`WithNodeSelectorFor` seems to have a usage in multiple locations in the test suite (although it is not yet directly used)and seems to be generic enough to reside in libvmi directly.


`WithClientPassthrough`, `WithSerialBios` seems like one-off options. We would just make a new pile of stuff(similar to utils file which is being witch-hunted) if we would be adding these options here.


@Barakmor1 
I think we can look for an analogy with functions. You don't make functions public just for the sake we might want to use them in the future. You make them private and you can make them public if the needs arise.

Comment (dankenigsberg) [reply tests/libvmi/vmi.go:165]: > If libvmi is supposed to fit many use cases and many users I don't see why not have a "pile of stuff" as long as it's generic and not coupled to a specific use-case. If we implement such an option that probably would fit external users - why make them reinvent the same wheel?

The other side of this is that typically, it is hard to define a good API for many use cases when you have only a single data point at hand. Hence it is prudent to keep things local until enough evidence show that the suggested API is good. In this case, it is not clear at all if something very specific as whether the VM bios speak over serial line should bother each reader of the main part of `libvmi` or should it sit in another file? If we do not know now what is the reasonable structure, we can postpone the decision to the day where the second user of the function shows up.

Comment (iholder-redhat) [reply tests/libvmi/vmi.go:165]: > The other side of this is that typically, it is hard to define a good API for many use cases when you have only a single data point at hand. Hence it is prudent to keep things local until enough evidence show that the suggested API is good. 

The way I see it, since we're in the position of designing a public library that is expected to be used by other users, but we have no such users yet, we have to take some sophisticated guesses. After all, there is no evidence for any API yet :)

> In this case, it is not clear at all if something very specific as whether the VM bios speak over serial line should bother each reader of the main part of `libvmi` or should it sit in another file?

I'm fine with keeping the main part of libvmi small, we could simply add separate files for specific configurations (e.g. firmware). That way we would keep the main file clean but still will provide many options for external users.

> If we do not know now what is the reasonable structure, we can postpone the decision to the day where the second user of the function shows up.

:+1: 

I think we should decide on a general approach moving forward. If we need to provide evidence for adding functions to libvmi, we should define what a "good evidence" is while we have no external users to guide us.

File: tests/libvmi/vmi.go
Comment (dankenigsberg) [reply tests/libvmi/vmi.go:150]: I attempted to provide an approximate definition: add a public only after you have two independent use cases for it.

It is very hard to predefine what a good API looked like. In healthy projects there is a tension between maintainers who want to keep a small bugless API and multiple users with conflicting requests to extend the API. A good API arises gradually from this tension.

File: tests/libvmi/vmi.go
Comment (dankenigsberg) [reply tests/libvmi/vmi.go:165]: I attempted to provide an approximate definition: add a public only after you have two independent use cases for it.

It is very hard to predefine what a good API looked like. In healthy projects there is a tension between maintainers who want to keep a small bugless API and multiple users with conflicting requests to extend the API. A good API arises gradually from this tension.

**Comment by Barakmor1** [comment]
@dankenigsberg @xpivarc
Removed the new Option and created local function instead.

File: tests/vmi_configuration_test.go
Comment (dankenigsberg) [inline tests/vmi_configuration_test.go:508]: when possible, please use static `func` definitions. They reduce mental burden. If they are well-named and placed at the bottom of the file, I do not need to read their content at all to understand the test.

**Review by dankenigsberg — CHANGES_REQUESTED** [review]

File: tests/libvmi/vmi.go
Comment (dankenigsberg) [inline tests/libvmi/vmi.go:140]: for the benefit of future git-loggers, please state why adding this function is a good idea (e.g we have multiple tests needing something like it).

nit: it feels better if placed next to `WithResourceMemory`.

File: tests/libvmi/vmi.go
Comment (iholder-redhat) [reply tests/libvmi/vmi.go]: > add a public API only after you have two independent use cases for it.

Again, in my eyes we are just another user. The ways we use these APIs should not set our design decisions imho.

> It is very hard to predefine what a good API looked like. In healthy projects there is a tension between maintainers who want to keep a small bugless API and multiple users with conflicting requests to extend the API. A good API arises gradually from this tension.

These options basically set a field into a struct. I don't think that there are much bugs to expose with such trivial implementation. And there are no users, so no conflicting requests, or any requests, exist. As said before, I think we would have to take guesses here. There is no other way imho. 

There's a very thin line between an internal library and a public library that is designed solely by how we are using it.

-----

Also, IMHO, this is a general discussion that needs to be done outside of the context of this PR. There are many other options (e.g. `WithSEV()`) that are already public. I think it would be much better to define what a "good API" is, what are the standards of adding a public option, etc, in a broader context.

If we change this in this PR only we end up in a mixed position and no clear path forward.

But all I can do is raise my protest :)

File: tests/vmi_configuration_test.go
Comment (Barakmor1) [reply tests/vmi_configuration_test.go]: Done.

File: tests/libvmi/vmi.go
Comment (Barakmor1) [reply tests/libvmi/vmi.go]: Done

File: tests/libvmi/vmi.go
Comment (Barakmor1) [reply tests/libvmi/vmi.go]: I think that we need a well defined convention that include this topic and also for the matter of static funcs definitions.
otherwise these rules will be violated / these Discussions will repeat them self  .

Comment (dankenigsberg) [reply tests/libvmi/vmi.go]: > As said before, I think we would have to take guesses here. There is no other way imho.

There is a way: keep the function local and do not commit on an API  until a second user shows up.

> But all I can do is raise my protest :)

No need for protest. If you feel sure enough that a certain function already belong in the library API, just say say so in the commit message and convince the library maintainer. It is a very thin line, and I don't think a clear-cut definition exists.

> also for the matter of static funcs definitions

Here, I believe, a clear-cut definition *does* exist. Lambdas are very cool and very powerful, but they incur cognitive burden and often too powerful. They force me to read their implementation when all I care is about their usage. Hence I strongly request to avoid them unless you really need their power (e.g need to change the function implementation on the fly).

Comment (Barakmor1) [reply tests/libvmi/vmi.go]: Still I think we can define conventions for usage of Lambdas (without enforcement).
I think that you have a point and the usage of Lambda in here was wrong but without
conventions this will not last as every reviewer have a different opinion
and there are many violations of the conclusions we talked about.

Comment (dankenigsberg) [reply tests/libvmi/vmi.go]: > Still I think we can define conventions for usage of Lambdas (without enforcement). I think that you have a point and the usage of Lambda in here was wrong but without conventions this will not last as every reviewer have a different opinion and there are many violations of the conclusions we talked about.

I am aware that we have many needlessly-confusing lambdas (I was just tipped about pile of them in [operator_test](https://github.com/kubevirt/kubevirt/blob/main/tests/operator_test.go#L113)). I think that a KubeVirt-wide convention against them would make a lot of sense.

File: tests/usbredir_test.go
Comment (dankenigsberg) [inline tests/usbredir_test.go:185]: there's nothing random about it. The name `vmi` would befit it better.
But more importantly, I think that this entire function can be inlined into its single user. `libvmi` makes it very clear to understand what is being built, so the function does not add much helpful abstraction.

File: tests/usbredir_test.go
Comment (dankenigsberg) [inline tests/usbredir_test.go:189]: using `tests.RunVMIAndExpectLaunch()` would remove much of this boilerplate.

**Review by dankenigsberg — CHANGES_REQUESTED** [review]

File: tests/usbredir_test.go
Comment (dankenigsberg) [inline tests/usbredir_test.go:26]: nit: nicer under `kubevirt.io/kubevirt/tests`

File: tests/usbredir_test.go
Comment (Barakmor1) [reply tests/usbredir_test.go:189]: Yes but don't we want to delete tests/utils.go ?
This will increase the coupling to this file.

File: tests/usbredir_test.go
Comment (Barakmor1) [reply tests/usbredir_test.go:185]: Right probably leftovers from the old code 
will update soon :)

Comment (Barakmor1) [reply tests/usbredir_test.go]: Done

File: tests/vmi_configuration_test.go
Comment (dankenigsberg) [inline tests/vmi_configuration_test.go:559]: The name of the function is good enough, I find this comment unhelpful. It only distracts the reader.

```suggestion
```

**Review by dankenigsberg — CHANGES_REQUESTED** [review]
Thanks. Feel free to merge despite these minor nits.

File: tests/libvmi/vmi.go
Comment (iholder-redhat) [reply tests/libvmi/vmi.go]: > > As said before, I think we would have to take guesses here. There is no other way imho.
> 
> There is a way: keep the function local and do not commit on an API until a second user shows up.

So we shouldn't expose any options since there is no second user for any of them?
 
> > But all I can do is raise my protest :)
> 
> No need for protest. If you feel sure enough that a certain function already belong in the library API, just say say so in the commit message and convince the library maintainer. It is a very thin line, and I don't think a clear-cut definition exists.

Right, but which factors make the "library maintainer" convinced?
I really think we should discuss and define these terms.

Right now I do think that the functions presented in this PR are general enough to be exposed. So I express it and try to convince. But it's hard to do so when we don't have any clear policy / guidelines or path forward.

Comment (iholder-redhat) [reply tests/libvmi/vmi.go]: > Here, I believe, a clear-cut definition does exist. Lambdas are very cool and very powerful, but they incur cognitive burden and often too powerful. They force me to read their implementation when all I care is about their usage. Hence I strongly request to avoid them unless you really need their power (e.g need to change the function implementation on the fly).

BTW, I don't agree with this either :)

I don't understand why a lambda functions "force me to read their implementation". Who's forcing you? :smile: Why are you forced to read the implementation of lambdas and not regular functions? Just skip the implementation and go straight to the usage, similar to any other function.

Defining a helper-lambda-function within another function lets you bound the context and scope of the function. This usually makes sense IMO when the function is meaningless outside of that context, which is pretty common for helper functions.

That being said, the fact that we have a thousand lambdas per test in situations that make no sense is clearly a bad habit.

Comment (iholder-redhat) [reply tests/libvmi/vmi.go]: > The difference is that lambda are defined in-place. Good tests are short. REALLY short. You glance at them and you know what they are doing. When I see a test that is calling a function named withSerialBIOS() that's all I need to know. If the function is defined in-scope, it hides more important things. The most important purpose of functions is to define an abstraction, which makes us think faster about the important things. This is challenged is the abstracted complexity is pushed to the face of the reader.

Let me give an example and let's see your opinion on it :)

```go
Context("VM with specific firmware", func() {

    setInvalidFirmware := func() {...}
    moreVeryContextSpecificHelperFunc := func() {...}

    It("ensure invalid firmware is rejected, func() {
        vm := ...
        setInvalidFirmware(vm)
        err := virtClient.Create(vm)
        Expect(err).To(HaveOccurred())
    })

})
```

In my opinion this usage is completely valid and encouraged. The advantages are:
- isFirmwareValid() has no meaning outside of this test. No one needs to set an invalid firmware outside of this specific context. Moreover, it would be very confusing to expose this function, even as a "private" function.
- The test is still very readable and straight forward.

Do you agree?

Comment (dankenigsberg) [reply tests/libvmi/vmi.go]: Sorry @Barakmor1 for hijacking your PR. Since we agree that your new functions are meaningful in the file context, we can take our discussion elsewhere.

> Do you agree?

I'm not sure that I do. But I'm willing to rescind my statement about the case against lamdas being "clear cut". There are exceptions.

I see no awkwardness in a file-local function named `withInvalidFirmware` and I find it hard to image what `moreVeryContextSpecificHelperFunc` may be. When the function implementation is hidden as `{...}` it does not distract me as much when I read the including context. I think that we can make better progress if you show me a real-life kubevirt-code named lambda that does not change, does not use context variable, yet improves readability. I suspect that there are such functions, but I bet we would find them in too-long too-cluttered files.

Comment (iholder-redhat) [reply tests/libvmi/vmi.go]: > Sorry @Barakmor1 for hijacking your PR. Since we agree that your new functions are meaningful in the file context, we can take our discussion elsewhere.

sounds good :+1: 
This is indeed beyond the scope of this PR :)

> 
> > Do you agree?
> 
> I'm not sure that I do. But I'm willing to rescind my statement about the case against lamdas being "clear cut". There are exceptions.
> 
> I see no awkwardness in a file-local function named `withInvalidFirmware` and I find it hard to image what `moreVeryContextSpecificHelperFunc` may be. When the function implementation is hidden as `{...}` it does not distract me as much when I read the including context. I think that we can make better progress if you show me a real-life kubevirt-code named lambda that does not change, does not use context variable, yet improves readability. I suspect that there are such functions, but I bet we would find them in too-long too-cluttered files.

I'm up for the challenge :)
And as said before, I think that you're right for most cases :) It's a great rule of thumb, but like all rules of thumb, they aren't correct for every situation imho

p.s. a modern IDE can "collapse" function implementations. So in my eyes these functions are always hidden behind `{...}`.

File: tests/usbredir_test.go
Comment (Barakmor1) [reply tests/usbredir_test.go]: make generate keep locating this here for some reason :\

Comment (dankenigsberg) [reply tests/usbredir_test.go]: it is creating it there; but if you move it elsewhere I bet it would not object. It's a bit simpler for humans to read when it sits among related files.

File: tests/vmi_configuration_test.go
Comment (Barakmor1) [reply tests/vmi_configuration_test.go]: Done.

File: tests/usbredir_test.go
Comment (Barakmor1) [reply tests/usbredir_test.go]: I think that there is a lane that fails if `make generate` change files but let's try this

File: tests/junit.functest.xml
Comment (dankenigsberg) [inline tests/junit.functest.xml:1]: unintentional, I presume?

**Review by dankenigsberg — APPROVED** [review]

**Review by enp0s3 — APPROVED** [review]
/approve

**Comment by Barakmor1** [comment]
Hey @dankenigsberg ,
I had to rebase.
Can you have another look please ?

File: tests/usbredir_test.go
Comment (xpivarc) [inline tests/usbredir_test.go:177]: Is this used?

File: tests/usbredir_test.go
Comment (xpivarc) [inline tests/usbredir_test.go:69]: This seems wrong.

File: tests/usbredir_test.go
Comment (Barakmor1) [reply tests/usbredir_test.go:177]: Yes in line 84.

Comment (xpivarc) [reply tests/usbredir_test.go:177]: Thanks, I see it now. Github failed me.

File: tests/usbredir_test.go
Comment (Barakmor1) [reply tests/usbredir_test.go:69]: Can you elaborate Please?
we had a discussion about the 32Mi:
https://github.com/kubevirt/kubevirt/pull/8075#discussion_r919005138

Comment (xpivarc) [reply tests/usbredir_test.go:69]: Everythin is alright. I missed the false `vmi, err = createVMI(virtClient, false)`. I apologize

Comment (brybacki) [reply tests/usbredir_test.go:69]: What do you think about making this 32Mi a named constant. A constant that we can also comment.

Another person will come here, and try to improve, and will be asking why 32 Mi and not 1Mi or 128Mi?

Comment (brybacki) [inline tests/usbredir_test.go:69]: What do you think about making this 32Mi a named constant. A constant that we can also comment.

Another person will come here, and try to improve, and will be asking why 32 Mi and not 1Mi or 128Mi?

Comment (Barakmor1) [reply tests/usbredir_test.go:69]: We had a discussion about the 32Mi:
https://github.com/kubevirt/kubevirt/pull/8075#discussion_r919005138

I definitely  agree that we should a follow up PR for `libvmi.NewDIskLess()` with the right default .
I plan to work on it soon.
But in my opinion this is out of the scope of this PR .

Comment (maiqueb) [reply tests/usbredir_test.go:69]: The thing is a well named constant is self-explanatory. 

Please consider the amount of digging required to find that comment when you're browsing the code (at least git blame to find out which commit - then find out which PR - then browse PR (hopefully the comment is not resolved ...). It's magnitudes harder than reading what it is from a const.

... lazy people will just roll with it. 

I would encourage you to make the reader's life easier.

Comment (Barakmor1) [reply tests/usbredir_test.go]: That's a good point i added a constant for now.

Comment (brybacki) [reply tests/usbredir_test.go]: now it looks better

**Pull request merged**
